File: madameautomata-bot-python-autoreg/bot/commands/collar.py

import discord
from discord import app_commands, Interaction
from ..theme import gothic_embed
from ..util.perm import is_domme, is_staff, is_sub
from ..db import Session, User, Collaring, Cage, JailSession, Star, Task, Worship, Config, CollarType, StarStatus
from ..util.time import parse_duration, random_case
from ..util.ids import cuid
from datetime import datetime, timedelta
from sqlalchemy import select, func, update, and_

def setup(bot):
    @bot.tree.command(name="collar", description="Collar a submissive with consent flow.")
    @app_commands.describe(sub="Submissive")
    async def collar(i: Interaction, sub: discord.User):
        member = i.guild.get_member(i.user.id)
        if not await is_domme(member):
            return await i.response.send_message("Only Dommes can use this.", ephemeral=True)

        view = discord.ui.View(timeout=120)
        select = discord.ui.Select(placeholder="Select collar type", options=[
            discord.SelectOption(label="Consideration Collar", value="CONSIDERATION", description="A trial period—serious, but not permanent."),
            discord.SelectOption(label="Training Collar", value="TRAINING", description="Active shaping and obedience expectations."),
            discord.SelectOption(label="Permanent Collar", value="PERMANENT", description="Lifetime devotion and duty."),
        ])
        choice = {"value": None}

        async def on_select(interaction: discord.Interaction):
            choice["value"] = interaction.data["values"][0]
            await interaction.response.defer()

        select.callback = on_select
        view.add_item(select)
        await i.response.send_message(embed=gothic_embed("Collar Selection", f"Choose a collar for **<@{sub.id}>**."), view=view)
        timed_out = await view.wait()
        if not choice["value"]:
            return

        # Consent buttons
        accept_view = discord.ui.View(timeout=120)
        async def accept_cb(interaction: discord.Interaction):
            if interaction.user.id != sub.id:
                return await interaction.response.send_message("Only the sub can accept/decline.", ephemeral=True)
            async with Session() as s:
                # upsert users
                u = await s.get(User, sub.id)
                if not u:
                    u = User(id=sub.id, is_sub=True)
                    s.add(u)
                else:
                    u.is_sub = True
                d = await s.get(User, i.user.id)
                if not d:
                    d = User(id=i.user.id, is_domme=True)
                    s.add(d)
                else:
                    d.is_domme = True
                s.add(Collaring(id=cuid(), domme_id=i.user.id, sub_id=sub.id, type=CollarType(choice["value"]), active=True))
                await s.commit()
            await interaction.response.edit_message(embed=gothic_embed("Collar Accepted", f"<@{sub.id}> is now collared by <@{i.user.id}>."), view=None)

        async def decline_cb(interaction: discord.Interaction):
            if interaction.user.id != sub.id:
                return await interaction.response.send_message("Only the sub can accept/decline.", ephemeral=True)
            await interaction.response.edit_message(embed=gothic_embed("Collar Declined", f"<@{sub.id}> declined the collar."), view=None)

        accept_view.add_item(discord.ui.Button(label="Accept", style=discord.ButtonStyle.success, custom_id="accept"))
        accept_view.add_item(discord.ui.Button(label="Decline", style=discord.ButtonStyle.danger, custom_id="decline"))

        # map callbacks
        for item in accept_view.children:
            if isinstance(item, discord.ui.Button) and item.custom_id=="accept":
                item.callback = accept_cb
            if isinstance(item, discord.ui.Button) and item.custom_id=="decline":
                item.callback = decline_cb

        await i.followup.send(embed=gothic_embed("Consent Required", f"**<@{sub.id}>**, do you accept the **{choice['value'].title()}** collar from **<@{i.user.id}>**?"), view=accept_view)

File: madameautomata-js-bot/src/commands/collar.js
import {
  SlashCommandBuilder,
  StringSelectMenuBuilder,
  ActionRowBuilder,
  ButtonBuilder,
  ButtonStyle
} from "discord.js";
import { prisma } from "../lib/prisma.js";
import { gothicEmbed } from "../theme/theme.js";
import { isDomme } from "../util/perm.js";
import { cuid } from "../util/ids.js";

const cmd = {
  data: new SlashCommandBuilder()
    .setName("collar")
    .setDescription("Collar a submissive with consent (type select + Accept/Decline).")
    .addUserOption(o => o.setName("sub").setDescription("Submissive").setRequired(true)),

  async run(i) {
    if (!await isDomme(i.member)) {
      return i.reply({ content: "Only Dommes can use this.", ephemeral: true });
    }
    const sub = i.options.getUser("sub", true);

    // 1) Type select
    const menu = new StringSelectMenuBuilder()
      .setCustomId(`collar-type:${i.id}`)
      .setPlaceholder("Select collar type")
      .addOptions(
        { label: "Consideration Collar", value: "CONSIDERATION", description: "Trial, but serious." },
        { label: "Training Collar", value: "TRAINING", description: "Active shaping & obedience." },
        { label: "Permanent Collar", value: "PERMANENT", description: "Lifetime devotion." }
      );
    const row = new ActionRowBuilder().addComponents(menu);

    await i.reply({
      embeds: [gothicEmbed("Collar Selection", `Choose a collar for **<@${sub.id}>**.`)],
      components: [row]
    });

    const msg = await i.fetchReply();
    const picked = await msg.awaitMessageComponent({
      time: 120000,
      filter: (x) => x.user.id === i.user.id && x.customId.startsWith("collar-type:")
    }).catch(() => null);

    if (!picked) return; // leave message as-is

    const collarValue = picked.values[0];
    await picked.update({
      embeds: [gothicEmbed("Consent Required", `**<@${sub.id}>**, do you accept the **${collarValue}** collar from **<@${i.user.id}>**?`)],
      components: []
    });

    // 2) Consent buttons
    const acceptBtn = new ButtonBuilder().setCustomId(`collar-accept:${i.id}`).setLabel("Accept").setStyle(ButtonStyle.Success);
    const declineBtn = new ButtonBuilder().setCustomId(`collar-decline:${i.id}`).setLabel("Decline").setStyle(ButtonStyle.Danger);
    const row2 = new ActionRowBuilder().addComponents(acceptBtn, declineBtn);

    const consentMsg = await i.followUp({ components: [row2] });

    const pressed = await consentMsg.awaitMessageComponent({
      time: 120000,
      filter: (x) =>
        x.user.id === sub.id &&
        (x.customId === `collar-accept:${i.id}` || x.customId === `collar-decline:${i.id}`)
    }).catch(() => null);

    if (!pressed) {
      return consentMsg.edit({
        embeds: [gothicEmbed("Consent Timeout", "No response from the sub.")],
        components: []
      });
    }

    if (pressed.customId.startsWith("collar-accept")) {
      await prisma.user.upsert({ where: { id: sub.id }, update: { isSub: true }, create: { id: sub.id, isSub: true } });
      await prisma.user.upsert({ where: { id: i.user.id }, update: { isDomme: true }, create: { id: i.user.id, isDomme: true } });
      await prisma.collaring.create({ data: { id: cuid(), dommeId: i.user.id, subId: sub.id, type: collarValue, active: true } });

      return pressed.update({
        embeds: [gothicEmbed("Collar Accepted", `<@${sub.id}> is now collared by <@${i.user.id}> (type: ${collarValue}).`)],
        components: []
      });
    } else {
      return pressed.update({
        embeds: [gothicEmbed("Collar Declined", `<@${sub.id}> declined the collar from <@${i.user.id}>.`)],
        components: []
      });
    }
  }
};

export default cmd;
